### DISCUSSION TITLE
Optimizing fetch cache for user-specific data in Next.js: Best Practices?

### DISCUSSION THREAD

**Discussion by lorenzo-polaris** [vercel/next.js#59527] [App Router]
Hello everyone,

I'm working on a Next.js project with a private user area and I'm contemplating the best approach to cache server-side requests for authenticated user data.

Here's a scenario for clarity:

**Public Access**: Non-authenticated users fetch articles, caching them with an "articles" tag.
**Editor Action**: When an editor publishes a new article, we invalidate the "articles" tag using revalidateTag, prompting a fresh fetch for non-authenticated users.

Now, consider adding a subscription feature:

**Authenticated Access**: Users fetch their subscriptions, caching them with a "subscriptions" tag.
**Subscription Update**: A user modifies their subscription, invalidating the "subscriptions" tag. This action causes the cache to become stale for all users, leading to unnecessary data refetching.

A potential solution is to include the user ID in the cache tag (e.g., "subscriptions/<user_id>"), but this raises concerns about cache management and data clearing on the Next.js server. Specifically:

- How efficiently can Next.js handle numerous user-specific cache tags?
- Is it possible to combine tag-based and time-based revalidation for optimal performance?

I'm eager to hear your thoughts and suggestions on managing cache for user-specific data in Next.js, particularly concerning scalability and efficiency.

Thank you!

**Comment by feliche93** [comment]
@lorenzo-polaris had the same question!

**Comment by lorenzo-polaris** [comment]
@feliche93 after some research and experiments, I found that user specific data should not be cached at all. 

This is also the default behavior of Next, if during your fetch request a special function is invoked (cookies() or headers() to get, for instance, user access token) then Next will not cache your request.

Hope it helps, if anyone is using a better approach please share!

**Comment by nick-barth** [comment]
I have a route that has 6 API calls on them, all for pretty massive lists.  The items that are returned in these lists, each have user data attatched to them, depending on their bearer token.  

How can I call, serve the 6 calls, on the server, so they are quick & SEO ready.  However, once the paged is served, I'd like to "hydrate" them with the user data, how is that done?

**Comment by lorenzo-polaris** [comment]
@nick-barth I had the same issue in my current project . My approach would be to:

- Fetch the public part of your data in an API call without any bearer token and display it in a component. This resource will be cached.
- Then, fetch the user specific data in an API call with the bearer token and "hydrate" your page/component. This can be done client side or in a server component (if different from the component that displays the public data).

I don't know if it solves your issue and if you have access to the backend. If you want to share more info I would be happy to help!

**Reply by nick-barth** [reply]
> Then, fetch the user specific data in an API call with the bearer token and "hydrate" your page/component.

So RSC grabs the lists, pass them to client components, that render the original lists, while int he background fetching lists with user data, and just bulldoze them with setState?  Is that what you mean? 

It just seems wasteful and slow, as we will now need 6x2 calls, 6 of which will be on the client.

**Reply by lorenzo-polaris** [reply]
Yes it is, but the initial 6 calls are still cached and performed once (hopefully at build time).  

How is your data displayed on the page? In my project I can separate the components that use the public data from the components that are user specific (within the same page).

It allows me to have the desired behavior (caching of the public API calls) and use RSCs without any client component. Even if both components make the same API call (one with token, one without) the public one is cached and does not impact performance.

If you want can't split the page/components and you just want a fast "fallback page" for search engines, I think that Next does not support that out of the box and you need to do implement this ugly logic.

If you find any better solution, I would be grateful if you can share that! Thank you.

**Comment by nick-barth** [comment]
Yes, this was also my conclusion.  I have no easy seperation.  I'll let you know how it preforms.

**Comment by williamlmao** [comment]
@leerob just got done watching your caching + email client videos - my biggest question in watching both of them was how caching works in relation to authenticated users. 

Reading this thread, it sounds like that's actually not what caching is meant for at all, but that seem to contradict you building an email client (although i get that might have just been an example). 

Just wanted to confirm - is there any way I can use the app router cache to cache data for a specific user? That would be really awesome if that was possible. I'm imagining passing in a userId into the revalidate tags. This would mean I wouldn't need to bring in a separate KV store.
